### ISSUE TITLE
Possible BUG may make your code not achieve the best performance. (background noise, etc.)

### ISSUE THREAD

**Issue by dathudeptrai** [kan-bayashi/ParallelWaveGAN#131]
Hi @kan-bayashi , there is a bug i think that make ur code cann't achieve the best performance for both melgan and PWG. That is after training 1 step generator, you should re-compute y_ then use this y_ for discriminator, but seem ur code not do that. In my experiment, re-compute y_ is crucial for obtain best quality. I have my tensorflow code for melgan, i can get the same performance as ur code but just around 2M steps from scratch (don't need PWG auxiliary loss to help convergence speed), but if i don't recompute y_, my tf code at 2M steps not good as 2M steps when recompute y_

**Comment by kan-bayashi** [comment]
Thank you for your suggestions!
I'm not sure which is the standard for GAN training, but it is better to change the update manner since your experiments shows better performance.
I will make PR.

**Comment by kan-bayashi** [comment]
BTW, how was the final quality? Is it improved? or just related to convergence speed?

**Comment by dathudeptrai** [comment]
I think that make sense, discriminator should use newest y_ for training, that make discriminator penalize generator better. U can ref two implementation on github (https://github.com/seungwonpark/melgan/blob/master/utils/train.py/#L85) and official code (https://github.com/descriptinc/melgan-neurips/blob/master/scripts/train.py/#L156). Note that official code training discriminator before generator but it's still re-compute D-fake for training generator. The quality on my language (vietnamese) is improve significantly. I'm training Ljspeech now on melgan only, i have 1M4 val audio here. 
[1460000steps.zip](https://github.com/kan-bayashi/ParallelWaveGAN/files/4526567/1460000steps.zip)

**Comment by kan-bayashi** [comment]
Thank you for sharing samples!
It seems very good :) I will follow your suggested manner!

**Comment by dathudeptrai** [comment]
BTW, i'm creating TF framework for Speech, now focus TTS :)), just like ur ESPNet :D. TF now training faster than pytorch with exactly parameter, batch_size, batch_max_step, the improvement on inference speed you knew before: D.

**Comment by kan-bayashi** [comment]
That sounds nice :)
But I love pytorch so I want facebook to make it faster 🚀

**Comment by dathudeptrai** [comment]
haha BTW, what is the license of this repo and ur ESPNET :)), i cann't retrain anything, there are many model and dataset, i plan to provide some training script, pretrained on my TF framework and the rest is convert from ur pytorch pretrained then support for inference only :D. Can i do that ?.

**Comment by kan-bayashi** [comment]
That's very cool. 
This repository is MIT license and ESPnet is Apache 2.0, so you can do it!

**Comment by dathudeptrai** [comment]
ok that nice, i will make TF models and pytorch models synchronized and use the same preprocesing here so atleast user can use pytorch for training and TF for inference :D. Again, thank for ur hard work :)).

**Comment by npuichigo** [comment]
@kan-bayashi Have you retrained models to verify this bug? It seems that after fixing this bug, the samples generated by pwavegan have less energy than reference recordings.
![image](https://user-images.githubusercontent.com/11533479/83370067-d6d54100-a3f0-11ea-9e7c-5ea46c276e4e.png)
